XmlLayout - Render LogEventInfo.Properties as XML

snakefoot opened the pull request:
Ex. SQL XML Column

```xml
<properties>
  <property key="Action">GetUsers</property>
  <property key="Controller">UserController</property>
</properties>
```

```sql
SELECT 	[Message]
  , [TimeStamp]
  , [Exception]
  , [Properties].value('(//property[@key="Action"]/node())[1]', 'nvarchar(max)') as Action
FROM [Logs]
WHERE [Properties].value('(//property[@key="Controller"]/node())[1]', 'nvarchar(max)') = 'UserController'
```

Ex. SMTP mails with advanced XHTML-formatting (colors etc):

`msg.BodyFormat = MailFormat.Html;`

snakefoot commented:
Still work-in-progress

304NotModified commented:
nice!

snakefoot commented:
Now I think it is ready for review.

304NotModified commented on src/NLog/Layouts/XmlAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+        /// Determines wether or not this attribute will be Json encoded.
+        /// </summary>
+        /// <docgen category='XML Attribute Options' order='100' />
+        public bool Encode
```
maybe default it to true, as 2 of the 3 ctors set it to true

snakefoot replied:
The default is true, as it gets the default value from `XmlEncodeLayoutRendererWrapper`

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        /// <docgen category='LogEvent Properties XML Options' order='10' />
+        public string PropertiesNodeName { get { return _propertiesNodeName; } set { _propertiesNodeName = value; _propertiesNodeNameFormat = value?.IndexOf('{') >= 0; } }
+        private string _propertiesNodeName = "property";
+        private bool _propertiesNodeNameFormat;
```
proposal: propertiesNodeNameHasFormat

snakefoot replied:
Changed

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        /// </summary>
+        /// <param name="nodeName">The name of the top XML node</param>
+        /// <param name="nodeValue">The value of the top XML node</param>
+        public XmlLayout(string nodeName, Layout nodeValue)
```
please move the ctors above the properties, thx!

snakefoot replied:
Changed

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        {
+            int orgLength = target.Length;
+            RenderXmlFormattedMessage(logEvent, target);
+            if (target.Length == orgLength && IncludeEmptyValue && !string.IsNullOrEmpty(NodeName))
```
if `IncludeEmptyValue == true`, should't the whole `if` result in `true` if `nodeName == ""`?

snakefoot replied:
Not sure I understand. It outputs the NodeName, even if nothing to output. If no NodeName, what should it then output?

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        [DefaultValue("logevent")]
+        public string NodeName { get => _nodename; set => _nodename = XmlHelper.XmlConvertToElementName(value?.Trim(), true); }
```
I think it should be RootElementName.

IMO XML consists of elements and attributes  (and yes, it's a tree structure with nodes, but that's not the terminology used with XML) - .e.g https://en.wikipedia.org/wiki/XML#Key_terminology

So I guess "Node" here is "Element" and I prefer to rename them all

snakefoot replied:
Changed

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        /// Value inside the top level XML element
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        public Layout NodeValue { get => NodeValueWrapper.Inner; set => NodeValueWrapper.Inner = value; }
```
proposal: ElementValue

snakefoot replied:
Changed

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,504 @@
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        [ArrayParameter(typeof(XmlLayout), "node")]
+        public IList<XmlLayout> Nodes { get; private set; }
```
Proposal: Elements

snakefoot replied:
Changed

snakefoot commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,476 @@
+    [Layout("XmlLayout")]
+    [ThreadAgnostic]
+    [ThreadSafe]
+    public class XmlLayout : Layout
```
Just like JsonLayout, then XmlLayout doesn't include IncludeNdc and IncludeNdlc

snakefoot commented on src/NLog/LayoutRenderers/Wrappers/XmlEncodeLayoutRendererWrapper.cs:
```diff
@@ -65,55 +63,44 @@ public XmlEncodeLayoutRendererWrapper()
 
         /// <summary>
-        /// Post-processes the rendered message. 
+        /// Gets or sets a value indicating whether to tranform newlines (\r\n) into (&#13;&#10;)
```
Don't like CDATA, prefer to escape/encode as needed.

snakefoot replied:
You don't need to excape `&` in `&amp;`

snakefoot commented on src/NLog/LayoutRenderers/Wrappers/XmlEncodeLayoutRendererWrapper.cs:
```diff
@@ -65,55 +63,44 @@ public XmlEncodeLayoutRendererWrapper()
+            for (int i = 0; i < target.Length; ++i)
             {
-                switch (text[i])
+                switch (target[i])
```
It is a performance feature. It validates the input StringBuilder, and if nothing to encode, then it just leaves it untouched. If something needs encoding then it converts to a temporary string, that is encoded after truncate of StringBuilder.

snakefoot commented on src/NLog/Internal/XmlHelper.cs:
```diff
@@ -117,16 +174,118 @@ internal static string XmlConvertToStringSafe(object value)
+                {
+                    if (i == 0 && (chr == 'x' || chr == 'X') && xmlElementName.Length >= 3)
+                    {
+                        if (char.ToLowerInvariant(xmlElementName[1]) == 'm' && char.ToLowerInvariant(xmlElementName[2]) == 'l')
```
There is a very nice comment at the top explaining the rules for XML-element-names.

304NotModified replied:
Could we add an unit test for this?

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+            RenderXmlFormattedMessage(logEvent, target);
+            if (target.Length == orgLength && IncludeEmptyValue && !string.IsNullOrEmpty(ElementName))
+            {
+                target.Append('<');
```
any idea if we could add an unit test for this case?

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+                    int beforeAttribLength = sb.Length;
+                    if (!RenderAppendXmlAttributeValue(attrib, logEvent, sb, sb.Length == orgLength))
+                    {
+                        sb.Length = beforeAttribLength;
```
any idea if we could add an unit test for this case?

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+            string propNameElement = null;
+            if (_propertiesElementNameHasFormat)
+            {
+                propNameElement = XmlHelper.XmlConvertToStringSafe(propName);
```
any idea if we could add an unit test for this case?

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+            }
+            else
+            {
+                sb.Append('>');
```
any idea if we could add an unit test for this case?

304NotModified commented on src/NLog/Internal/XmlHelper.cs:
```diff
@@ -117,16 +174,118 @@ internal static string XmlConvertToStringSafe(object value)
+                {
+                    if (chr == '_')
+                    {
+                        sb?.Append(chr);
```
Could we add an unit test for this?

304NotModified commented on src/NLog/Internal/XmlHelper.cs:
```diff
@@ -117,16 +174,118 @@ internal static string XmlConvertToStringSafe(object value)
+                    sb = new StringBuilder(xmlElementName.Length);
+                    if (i > 1)
+                        sb.Append(xmlElementName, 0, i - 1);
+                    if (i == 0 && char.IsWhiteSpace(chr))
```
Could we add an unit test for this?

snakefoot commented:
You can have as many unit-test that you like, just a matter of work-hours.

Like to agree on API before starting to lock things down with unit-tests. But happy you like it.

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        public string PropertiesElementKeyAttribute { get; set; } = "key";
+
+        /// <summary>
+        /// XML attribute format to use when rendering property-value
```
format = name here, isn't?

format sounds like formatstring

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Layouts/XmlAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        /// <summary>
+        /// Determines wether or not this attribute will be Json encoded.
```
json -> xml

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Layouts/XmlAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+    using NLog.Internal;
+
+    /// <summary>
+    /// JSON attribute.
```
json -> xml :)

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Layouts/XmlAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+        /// </summary>
+        /// <param name="name">The name of the attribute.</param>
+        /// <param name="layout">The layout of the attribute's value.</param>
+        /// <param name="encode">Encode value with json-encode</param>
```
json -> xml

snakefoot replied:
Fixed

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        /// Auto indent and create new lines
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        public bool IndentXml { get; set; }
```
maybe just "Indent"

304NotModified commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmlLayout"/> class.
+        /// </summary>
+        public XmlLayout()
```
maybe we need also "MaxRecursionLimit" in the near future

snakefoot replied:
There is no reflection happiness in this PR. It is like JsonLayout V1

304NotModified commented on src/NLog/Layouts/XmlAttribute.cs:
```diff
@@ -0,0 +1,111 @@
+        /// </summary>
+        /// <docgen category='XML Attribute Options' order='10' />
+        [RequiredParameter]
+        public Layout Layout
```
I see it's Layout in `JsonAttribute`

snakefoot commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        [DefaultValue("logevent")]
+        public string ElementName { get => _elementName; set => _elementName = XmlHelper.XmlConvertToElementName(value?.Trim(), true); }
```
Don't like root, because it usually means "one". Notice that you can perform nesting of XmlLayout (See unit-test with xml-example).

snakefoot commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        /// Value inside the top level XML element
+        /// </summary>
+        /// <docgen category='XML Options' order='10' />
+        public Layout ElementValue { get => _elementValueWrapper.Inner; set => _elementValueWrapper.Inner = value; }
```
Value is very vague. The value belongs to the Element.

snakefoot commented on src/NLog/Layouts/XmlLayout.cs:
```diff
@@ -0,0 +1,505 @@
+        /// Initializes a new instance of the <see cref="XmlLayout"/> class.
+        /// </summary>
+        public XmlLayout()
+            : this("logevent", null)
```
No idea. Just a default value.

304NotModified commented:
@snakefoot could you please resolve the merge conflict? Thanks!

304NotModified commented:
I think this one needs some unit tests, then it's good to go. I try to work on this when I find time for it.

304NotModified commented:
Hi @snakefoot 

I've added some tests. Some fails, could you please give your opinion? (see separate commits for every testcase)

snakefoot commented:
> @304NotModified I've added some tests. Some fails, could you please give your opinion?

I have added a commit that fixes the IncludeEmptyValue-test, and changed the render to better support mixed-mode-xml.

snakefoot commented:
Started working on https://github.com/NLog/NLog/wiki/XmlLayout

304NotModified commented:
@snakefoot  this could be helpful:

[XmlLayout.md.txt](https://github.com/NLog/NLog/files/2891479/XmlLayout.md.txt)

304NotModified commented:
updated the xlmlayout. Added some more sections, not sure if it's better. Feel free to restructure

304NotModified commented:
bit stange that attribute has "name" and element has "ElementName " 👼

snakefoot commented:
> bit stange that attribute has "name" and element has "ElementName "

Think it was your decision. Don't mind changing it.

304NotModified commented:
Then maybe is my mistake 😂

snakefoot commented:
Created #3137 as example of rename.

304NotModified commented:
@snakefoot I'm happy with the docs now, are you also?

One thing I'm doubting of,

the first "definition" is a bit strange:


![image](https://user-images.githubusercontent.com/5808377/53767510-59154380-3ed6-11e9-8bc2-a683e329c67e.png)


 It's now a combo of a definition (`includeAllProperties="`Boolean`" ) and an example. Not that I am a fan of the layout of the definitions of other targets... 

Should we convert it to an example or a definition?

snakefoot commented:
@304NotModified Made some updates to the Wiki-page for XmlLayout: https://github.com/NLog/NLog/wiki/XmlLayout
